### PR TITLE
Add delete metadata buttons

### LINKED ISSUE #4279: Cannot enter repeated keys in process creation
If several metadata entries for the same key have to be recorded when creating the process, this is not possible because there is no button to add further metadata fields of a type.

Buttons in metadata editor:

Buttons in metadata editor

No buttons when creating a process:

No buttons

**Goal:** It must be possible to record several metadata entries for the same key. To do this, the “plus” button must be displayed, as in the metadata editor. This must also be possible for nested keys.

### LINKED ISSUE #4284: Doubling nested metadata entries does not copy the values
Doubling nested metadata entries:

Entry to duplicate

The entry has been doubled, but without the data in it:

Second entry has no values

**Goal:** When duplicating nested entries, also duplicate the values.

### LINKED ISSUE #4457: When adding sub-field in Metadata-Editor, existing sub-fields are doubled
My ruleset has some complex keys/fields, that have sub-fields. E.g. personGroup or corporationGroup.

In Metadata-Editor, I can add such a key, e.g. corporationGroup - that is ok, group is added.

Now I want to add sub-fields: 
- First field (e.g. Displayform) is ok. 
- Next field (e.g. IdentifierPPN) can be added, but another field "DisplayForm" is also added

So, an already existing sub-field in metadata-editor gets doubled, when adding a new sub-field.

Update: this does NOT happen, when I first add the sub-fields one-by-one without entering values. 

The doubling only appears, if I am adding a sub-field, input a value, save - and then add the next sub-field.

Screenshot:
md-editor-corporation-field-doubling

### LINKED ISSUE #4468: Meta data editor - LABEL (mets-domain) key cannot be deleted but duplicated, even if it is "maxOccurs=1"
I have in my Rule Set this:

```
<key id="LABEL" domain="mets:div">
	<label>Main title ‹METS›</label>
	<label lang="de">Haupttitel ‹METS›</label>
</key>

<restriction division="Monograph" unspecified="forbidden">
	<permit key="LABEL" maxOccurs="1"/>

```
Let's assume via catalog import I get the LABEL set.
In Meta Data Editor it looks like this:
grafik
Which should not be the case (the duplicate button should be greyed-out).
Additionally, I cannot remove this meta data. (it stays in meta.xml, even if meta data editor UI and validation is ok).

My personal guess: Somehow "maxOccurs" is mixed up with "minOccurs".

Please check.

### LINKED ISSUE #4893: The buttons for duplicating metadata in the subfields of metadata groups is always disabled
The buttons for duplicating metadata in the subfields of metadata groups is always disabled while these metadata can be added by add metadata button ('+' button)

[
Bildschirmfoto 2021-12-22 um 20 44 39
](url)

Bildschirmfoto 2021-12-22 um 20 49 27

### PR DESCRIPTION
**Pull request by IkramMaalej**
Part of #4322 
Fixes #4279
Fixes #4457
Fixes #4468 
Fixes #4893 
Fixes #4284

This  PR :
- offers the possibility to add, duplicate or delete a metadata as in metadata editor  while creating a new process .
- fixes all above announced issues

### CODE REVIEW

**Comment by solth** [comment]
Does this fix #4279?

**Comment by solth** [comment]
I get a `PropertyNotFoundException` when deleting a metadata that I just added:

https://user-images.githubusercontent.com/19183925/147226550-58a8c650-8d69-49e8-bb23-0042d6c18632.mov

File: Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CreateProcessForm.java [hunk 1]
```diff
@@ -584,4 +596,56 @@ public int getProgress() {
...
+    /**
+     * Check and return whether the given ProcessDetail 'processDetail' is contained in the current list of addable
+     * metadata types in the addDocStrucTypeDialog.
+     *
+     * @param treeNode treeNode to be added
+     * @return whether the given ProcessDetail can be added or not
+     */
+    public boolean canBeAdded(TreeNode treeNode) throws InvalidMetadataValueException {
+        if (Objects.isNull(treeNode.getParent().getParent())) {
+            if (Objects.nonNull(processMetadata.getSelectedMetadataTreeNode()) || Objects.isNull(addMetadataDialog.getAddableMetadata())) {
```
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CreateProcessForm.java:609]: This line is probably too long

File: Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CreateProcessForm.java [hunk 2]
```diff
@@ -584,4 +596,56 @@ public int getProgress() {
...
+                this.addMetadataDialog.prepareAddableMetadataForStructure();
+            }
+        } else if (!Objects.equals(processMetadata.getSelectedMetadataTreeNode(), treeNode.getParent())
+                || Objects.isNull(addMetadataDialog.getAddableMetadata())) {
+            prepareAddableMetadataForGroup(treeNode.getParent());
+        }
+        if (Objects.nonNull(addMetadataDialog.getAddableMetadata())) {
+            return addMetadataDialog.getAddableMetadata().stream()
+                    .map(SelectItem::getValue).collect(Collectors.toList()).contains(((ProcessDetail) treeNode.getData()).getMetadataID());
```
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/forms/createprocess/CreateProcessForm.java:618]: This line is probably too long

File: Kitodo/src/main/java/org/kitodo/production/forms/createprocess/ProcessDetail.java [hunk 3]
```diff
@@ -166,4 +155,10 @@ public String getLabel() {
     public int getOccurrences() {
         return container.getOccurrences(this.getMetadataID());
     }
+
```
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/forms/createprocess/ProcessDetail.java:158]: Missing JavaDoc

File: Kitodo/src/main/java/org/kitodo/production/forms/createprocess/ProcessDetail.java [hunk 4]
```diff
@@ -122,6 +109,8 @@ public String getLabel() {
      *             if the metadata form contains syntactically wrong input
      */
     public abstract Collection<Metadata> getMetadata() throws InvalidMetadataValueException;
+    
+    public abstract Collection<Metadata> getMetadata(boolean skipEmpty) throws InvalidMetadataValueException;
```
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/forms/createprocess/ProcessDetail.java:113]: Missing JavaDoc

File: Kitodo/src/main/java/org/kitodo/production/forms/createprocess/ProcessMetadata.java [hunk 5]
```diff
@@ -89,4 +123,31 @@ public void preserve() {
...
+     *
+     * @return whether any further metadata can be added to currently selected structure element.
+     */
+    public boolean metadataAddableToStructureElement() throws InvalidMetadataValueException {
+        createProcessForm.getAddMetadataDialog().prepareAddableMetadataForStructure();
+        return !(createProcessForm.getAddMetadataDialog().getAddableMetadata().isEmpty());
+    }
+
+    /**
+     * Adds an empty table line with the given type.
```
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/forms/createprocess/ProcessMetadata.java:138]: What "given type"? This function does not appear to have parameters. Please update JavaDoc.

File: Kitodo/src/main/resources/messages/messages_de.properties [hunk 6]
```diff
@@ -580,6 +580,7 @@ messages=Nachrichten
 metadata=Metadaten
 metadataCorrupt=Die Metadatendatei ist fehlerhaft, bitte informieren Sie ihren Administrator.
 metadataDelete=Metadaten l\u00F6schen
+metadataDeleteNotAllowed=Metadaten ist in Regelsatz als 'Required' oder "alwaysShowing" konfiguriert
```
Comment (solth) [inline Kitodo/src/main/resources/messages/messages_de.properties:583]:
```suggestion
metadataDeleteNotAllowed=Metadatum ist in Regelsatz als 'Required' oder "alwaysShowing" konfiguriert
```

File: Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css [hunk 7]
```diff
@@ -1273,6 +1299,28 @@ Import form
     width: calc(100% - var(--default-double-size));
 }
 
+#editForm\:processFromTemplateTabView\:metadataTable tr.ui-widget-content:not([data-prk='root']) span > .ui-inputfield {
+    width: calc(50% - var(--default-double-size) - 93px);
+}
+
+#editForm\:processFromTemplateTabView\:metadataTable tr.ui-widget-content:not([data-prk='root']) span > .ui-inputfield.hasDatepicker {
+    width: calc(50% - 4 * var(--default-half-size) - 93px - 39px);
```
Comment (solth) [inline Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css:1307]:
```suggestion
    width: calc(50% - var(--default-double-size) - 132px);
```

File: Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css [hunk 8]
```diff
@@ -1273,6 +1299,28 @@ Import form
...
+}
+
+#editForm\:processFromTemplateTabView\:metadataTable tr.ui-widget-content span > .ui-inputfield.hasDatepicker {
+    width: calc(50% - var(--default-half-size) - 93px - 39px);
```
Comment (solth) [inline Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css:1311]:
```suggestion
    width: calc(50% - var(--default-half-size) - 132px);
```

**Review by solth — CHANGES_REQUESTED** [review]
Just a first part of the review. Will test more later.

One problem I encountered was that when I add a metadata group (for example "Person") during process creation and then add individutal metadata entries (for example "Display name" and "Family name") to the new group, the second entry will replace the first if I didn't add any values to the entries yet:

https://user-images.githubusercontent.com/19183925/148578362-f4c39f6b-783f-4212-8b83-0b4279745bdd.mov

File: Kitodo/src/main/resources/messages/messages_de.properties (hunk 6 above)
Comment (matthias-ronge) [reply Kitodo/src/main/resources/messages/messages_de.properties:583]: _Metadatum ist im Regelsatz vorgegeben_

**Comment by matthias-ronge** [comment]
Maybe this is because of `additionallySelectedFields` only contains the additionally selected field? When adding another sub-field, the metadata tree table is rebuilt and the first additionally selected field is no longer known. I think it is complicated to fix this

**Comment by andre-hohmann** [comment]
### Summary
- #4279 -> is fixed
- #4457 -> is fixed
- #4468 -> is fixed
- #4465 -> **i cannot assess this problem with our current configuration**
- #4893 -> is fixed
- #4284 -> **is not fixed - see review below**

### Review
1. _4284_: The metadata group can be copied with its values, but when the metadata editor is closed and opened again, the copied metadata group is lost. When the process is validated an exception occurs - see below. 
2. _Add metadata during creation of process_:  If a metadata field is imported and the same metadata field is added manually to the imported metadata, the following behaviour occurs: The field is added behind the existing metadata field, but contains the value of the metadata field, which has been at this position before. The values of all following metadata fields are mixed up in the same way. 
If a metadata field is added, which is not shown in the list of the imported metadata, everything works fine. 
The goal should be that the metadata fields and its values should not be mixed up when existing metadata fields are added manually. 

### Exception message 4284
```
An error has occurred:
Date/time: 2022-02-09 08:27:09
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36
User IP: 134.100.172.21
Request URI: /cePreview/pages/metadataEditor.jsf
Ajax request: Yes
Status code: 500
Exception type: class java.lang.NullPointerException
Exception message:
Exception UUID:
Stack trace:
java.lang.NullPointerException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
	at java.base/java.util.stream.FindOps$FindOp.evaluateParallel(FindOps.java:160)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.findAny(ReferencePipeline.java:548)
	at org.kitodo.dataeditor.ruleset.NestedKeyView.retrieveOrCompute(NestedKeyView.java:468)
	at org.kitodo.dataeditor.ruleset.NestedKeyView.lambda$storeValues$3(NestedKeyView.java:460)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at org.kitodo.dataeditor.ruleset.NestedKeyView.storeValues(NestedKeyView.java:459)
	at org.kitodo.dataeditor.ruleset.NestedKeyView.createAuxiliaryTable(NestedKeyView.java:304)
	at org.kitodo.dataeditor.ruleset.NestedKeyView.getSortedVisibleMetadata(NestedKeyView.java:422)
	at org.kitodo.validation.metadata.MetadataValidation.checkForMandatoryQuantitiesOfTheMetadataRecursive(MetadataValidation.java:263)
	at org.kitodo.validation.metadata.MetadataValidation.checkMetadataRules(MetadataValidation.java:172)
	at org.kitodo.validation.metadata.MetadataValidation.validate(MetadataValidation.java:123)
	at org.kitodo.production.services.validation.MetadataValidationService.validate(MetadataValidationService.java:181)
	at org.kitodo.production.forms.dataeditor.DataEditorForm.validate(DataEditorForm.java:409)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.el.parser.AstValue.invoke(AstValue.java:246)
	at org.apache.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:266)
	at org.jboss.weld.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
	at org.jboss.weld.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
	at org.jboss.weld.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
	at org.jboss.weld.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
	at org.apache.myfaces.view.facelets.el.ContextAwareTagMethodExpression.invoke(ContextAwareTagMethodExpression.java:96)
	at javax.faces.event.MethodExpressionActionListener.processAction(MethodExpressionActionListener.java:83)
	at javax.faces.event.ActionEvent.processListener(ActionEvent.java:58)
	at javax.faces.component.UIComponentBase.broadcast(UIComponentBase.java:429)
	at javax.faces.component.UICommand.broadcast(UICommand.java:103)
	at javax.faces.component.UIViewRoot._broadcastAll(UIViewRoot.java:1255)
	at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:420)
	at javax.faces.component.UIViewRoot._process(UIViewRoot.java:1741)
	at javax.faces.component.UIViewRoot.processApplication(UIViewRoot.java:935)
	at org.apache.myfaces.lifecycle.InvokeApplicationExecutor.execute(InvokeApplicationExecutor.java:42)
	at org.apache.myfaces.lifecycle.LifecycleImpl.executePhase(LifecycleImpl.java:195)
	at org.apache.myfaces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:142)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:204)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.kitodo.production.servletfilter.EncodingFilter.doFilter(EncodingFilter.java:69)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.kitodo.production.security.SecurityObjectAccessFilter.doFilter(SecurityObjectAccessFilter.java:86)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.ConcurrentSessionFilter.doFilter(ConcurrentSessionFilter.java:155)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:66)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:196)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:661)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:698)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:364)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:624)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:831)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1650)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException
	at org.kitodo.dataeditor.ruleset.NestedKeyView.lambda$retrieveOrCompute$4(NestedKeyView.java:468)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1631)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindTask.doLeaf(FindOps.java:319)
	at java.base/java.util.stream.AbstractShortCircuitTask.compute(AbstractShortCircuitTask.java:115)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

**Comment by IkramMaalej** [comment]
@andre-hohmann 
> 4284: The metadata group can be copied with its values, but when the metadata editor is closed and opened again, the copied metadata group is lost. 

Did you mean that the metadata editor is closed with 'Save & exit' button or with 'Exit' button?

**Comment by andre-hohmann** [comment]
I mean 'Save & exit'. 

I guess you ask, because it works for you. Therefore i tested a bit and found out that it only happens, when the metadata fields are only added and not changed. 
When i add at least one character in one sub-field of the duplicated metadata group, the whole metadata group is saved correctly and is shown after opening the metadata editor again.

**Comment by IkramMaalej** [comment]
@andre-hohmann 
> I mean 'Save & exit'.
> 
> I guess you ask, because it works for you. Therefore i tested a bit and found out that it only happens, when the metadata fields are only added and not changed. When i add at least one character in one sub-field of the duplicated metadata group, the whole metadata group is saved correctly and is shown after opening the metadata editor again.

What you have described above is the intended logic so far, because for storing the metadata  we are using an object that doesn't allow duplicates (2 identical objects). But if we now want to save duplicates, then we have to adjust the logic accordingly.

**Comment by andre-hohmann** [comment]
I do not need duplicate objects, but i wanted to describe my observation. Usually, it is a bad sign, if (meta)data disappears.  Maybe some users copy some elements and want them to adjust later, ...

In such cases, it would be helpful to show a message. In this case for example: 
- `The duplicate metadata object is not saved when the editor closed`. 

Then it is clear that the following behaviour is intended.

**Comment by matthias-ronge** [comment]
@IkramMaalej is right: The metadata is stored internally as a [set (in the mathematical sense)](https://en.wikipedia.org/wiki/Set_(mathematics)). A set is defined as that its elements have no order, and a set cannot contain the same element more than once. Java automatically applies these rules to the metadata. This means that all elements of the set are compared with each other, and if an element is identical to another, i.e. it occurs more than once, it is deduplicated. For metadata, that's absolutely fine if the same thing doesn't occur multiple times.

**The contract of the set does assure to you, that no metadata, that is not 100% identical, will ever be deleted.**

This is not a bug.

**Comment by andre-hohmann** [comment]
@matthias-ronge : Thanks a lot for clarification. Then #4284 is solved, too. I do not know how to change the previous comments, but maybe that is the way to make the transparent discussions. 

Thus, from my point of view, everything in the PR works fine except of:  
- #4465, which i cannot assess.

**Review by andre-hohmann — APPROVED** [review]
https://github.com/kitodo/kitodo-production/issues/4465 is not assessed.

**Pull request merged**